### ISSUE TITLE
Question about verify method

### ISSUE THREAD

**Issue by nashgao** [lcobucci/jwt#280]
If a user logs in and receives a token. The user can obtain the payload by decoding the token. What if the user modifies some content of the payload such as scope or something, does the verify method can detect that(which means return false)? 
From the demo it is:

var_dump($token->verify($signer, 'testing 1')); // false, because the key is different
var_dump($token->verify($signer, 'testing')); // true, because the key is the same

which means if someone tries to modify the payload, then the key will be changed? How can I get the changed key?

The way I'm thinking is that, after getting the token returned by the user, recalculate the signature based on header, payload and the key. Then compare this recalculated signature with the signature sent by the user, if they match then no modification of the payload and vice versa. 

Am I misunderstanding the verify method? Or maybe I'm just a bit confused.

**Comment by medmin** [comment]
I am also confused.

So, the server generates a token string and sends the token string to the client app ( such as a web app built with React ). And the client app sends this token string back to server.

But the `verify` method is the `Token` **Object**'s function. How do I verify a  token `string` ?

**Comment by lcobucci** [comment]
@nashgao @medmin let's try to address the confusion :smile: 

First thing we have to remember is that the server side can **never** ensure that a request came from a trustworthy client. We always have to validate things.

> If a user logs in and receives a token. The user can obtain the payload by decoding the token. What if the user modifies some content of the payload such as scope or something, does the verify method can detect that(which means return false)?
> From the demo it is:
> 
> ```php
> var_dump($token->verify($signer, 'testing 1')); // false, because the key is different
> var_dump($token->verify($signer, 'testing')); // true, because the key is the same
> ```
> 
> which means if someone tries to modify the payload, then the key will be changed? How can I get the > changed key?

A signed JWT is composed with three parts: headers, claims, and signature. The signature is [created by the **issuer**](https://github.com/lcobucci/jwt/blob/3.2/README.md#creating) using a key and when we (issuer or any client) verify the signature a compatible key must be used.

**Example**

Let's say that the result of `base64_url(json_encode($headers)) . '.' . base64_url(json_encode($claims))` is `A.B`. The signature is created passing that value to a hashing algorithm with a key, so: `fancy_hash('A.B', $key)`, which for our example is `C`. Our token here will then be `A.B.C`, which is sent to the client.

When receiving a request we'll first [parse the token](https://github.com/lcobucci/jwt/blob/3.2/README.md#parsing-from-strings) (which covers what you asked @medmin), then perform the [validations](https://github.com/lcobucci/jwt/blob/3.2/README.md#validating) and [verification of the signature](https://github.com/lcobucci/jwt/blob/3.2/README.md#token-signature).

The verification is done by checking that the hash created as signature was created with the expected algorithm and key based on the payload. If we receive something like `A.B.D` (signature doesn't match our expectation), `A.E.C` (modified claims with same signature), or any other possible variation we should **reject** the request.

The method `Token#verify()` is responsible for that check, so it fails (returns `false`) under all those conditions.

I hope this make things clearer :+1: 

I'll close this issue but please feel free to extend your question.

**Comment by medmin** [comment]
Thanks for the detailed explanation. I really appreciate it.

It's very clear. Thank you so much!

**Comment by nashgao** [comment]
Thank you for the reply, really appreciate it as well. 
The reason why I asked this question is that: I generated a token, then I only changed the payload and use the verify method to test it. Then I use this token to log in, the verify method didn't reject this token. 

Assume the header is A, the payload is B and signature is C for the original token.
After I change the scope in the payload, then the header is  A, payload should b B+, and the signature should be C. 
Under this circumstance, I expect the verify method to reject this changed token, but it didn't. That's why I was so confused. Probably there was some misuse at the beginning so it didn't reject the changed token. Anyway, I've made some changes on top of it and it works fine. 

Thanks again for your reply.
